### PR TITLE
[rush-lib] prevent workspace dependencies in decoupledLocalDependencies

### PR DESCRIPTION
**Pull request by jamesmaher-dd**
## Summary

Adds a check in install step to prevent a host project depending on a workspace: package version if the package is also listed in decoupledLocalDependencies.

Currently unaware of any use case that this would be intended. We came across this issue when migrating a dependency from an external feed (e.g. `"@<namespace>/packageA": "1.2.3"`) to a workspace dependency, and didn't remove the package from `decoupledLocalDependencies`. 

## Details

We came across an issue which allows a host package to depend on a package with `workspace:` notation while also listing that package in its `decoupledLocalDependencies`. Because this is currently not checked, this can result in a missing dependency in CI runs because the `package@workspace:` is not guaranteed to be built prior to the package that depends on it, so here we're adding a check for it.

## Replication steps

This is a very quick example to demonstrate how this situation occurs:
```
# hostProject1 package.json
"dependencies": {
  "@<namespace>/packageA": "workspace:*",
}
```

```
# rush.json
"projects": [
  {
    "packageName": "@<namespace>/hostProject1",
    "projectFolder": "apps/hostProject1",
    "decoupledLocalDependencies": ["@<namespace>/packageA"],
  },
```

## How it was tested

Tested with local rush against our repo by replicating with the steps above and running `rush update` or `rush install`, and no warning logged or error being thrown.

Adding this changeset against our repo with the same replication results in the error added in this PR being thrown, preventing the erroneous state.

### CODE REVIEW

**Review by aramissennyeydd — APPROVED** [review]
Nice!

**Review by justindeocampodd — APPROVED** [review]

**Comment by jamesmaher-dd** [comment]
@microsoft-github-policy-service agree [company="doordash"]

**Comment by jamesmaher-dd** [comment]
@microsoft-github-policy-service agree company="doordash"

File: common/changes/@microsoft/rush/jmaher-prevent-decoupled-workspace-deps_2024-07-30-15-46.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "prevent workspace deps in decoupledLocalDependencies",
```
Comment (iclanton) [inline common/changes/@microsoft/rush/jmaher-prevent-decoupled-workspace-deps_2024-07-30-15-46.json:4]:
```suggestion
      "comment": "Emit an error if a `workspace:` specifier is used in a dependency that is listed in `decoupledLocalDependencies`.",
```

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 2]
```diff
@@ -279,7 +279,25 @@ export class WorkspaceInstallManager extends BaseInstallManager {
...
+          rushProject.decoupledLocalDependencies.has(name)
+        ) {
+          // If the dependency is a local project that is decoupled, then we need to ensure that it is not specified
+          // as a workspace project. If it is, then we need to update the package.json to remove the workspace notation.
+          // eslint-disable-next-line no-console
+          console.log();
+          // eslint-disable-next-line no-console
+          console.log(
+            Colorize.red(
+              `"${rushProject.packageName}" depends on package ${name}@${version}, but also lists it in` +
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:293]:
```suggestion
              `"${rushProject.packageName}" depends on package ${name}@${version}, but also lists it in ` +
```

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 3]
```diff
@@ -279,7 +279,25 @@ export class WorkspaceInstallManager extends BaseInstallManager {
...
             continue;
           }
-        } else if (dependencySpecifier.specifierType === DependencySpecifierType.Workspace) {
+        } else if (
+          dependencySpecifier.specifierType === DependencySpecifierType.Workspace &&
+          rushProject.decoupledLocalDependencies.has(name)
+        ) {
+          // If the dependency is a local project that is decoupled, then we need to ensure that it is not specified
+          // as a workspace project. If it is, then we need to update the package.json to remove the workspace notation.
+          // eslint-disable-next-line no-console
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:288]: Use `this._terminal` instead.

**Review by iclanton — APPROVED** [review]

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 3 above)
Comment (jamesmaher-dd) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts]: Updated 👍🏻

**Comment by jamesmaher-dd** [comment]
hey @iclanton 👋🏻, how do we feel about merging this?

**Review by iclanton — APPROVED** [review]

**Pull request merged**